src/Agent.Worker/TaskRunner.cs

```diff
@@ -13,6 +15,67 @@
 
 namespace Microsoft.VisualStudio.Services.Agent.Worker
 {
+    [ServiceLocator(Default = typeof(SignatureService))]
+    public interface ISignatureService : IAgentService
+    {
+        Task<Boolean> VerifyAsync(Definition definition);
+    }
+
+    public class SignatureService : AgentService, ISignatureService
+    {
+        public async Task<Boolean> VerifyAsync(Definition definition)
+        {
+            // This is used for the Checkout task.
+            // We can consider it verified since it's embedded in the Agent code.
+            if (String.IsNullOrEmpty(definition.ZipPath))
+            {
+                return true;
+            }
+
+            // Find NuGet
+            String nugetPath = WhichUtil.Which("nuget", require: true);
+
+            var configurationStore = HostContext.GetService<IConfigurationStore>();
+            AgentSettings settings = configurationStore.GetSettings();
+            String fingerprint = settings.Fingerprint;
+
+            String taskZipPath = definition.ZipPath;
+            String taskNugetPath = definition.ZipPath.Replace(".zip", ".nupkg");
+
+            // Rename .zip to .nupkg
+            File.Move(taskZipPath, taskNugetPath);
+
+            String arguments = $"verify -Signatures \"{taskNugetPath}\" -CertificateFingerprint {fingerprint} -Verbosity Detailed";
+
+            // Run nuget verify
+            using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+            {
+                int exitCode = await processInvoker.ExecuteAsync(workingDirectory: HostContext.GetDirectory(WellKnownDirectory.Root),
+                                                                 fileName: nugetPath,
+                                                                 arguments: arguments,
+                                                                 environment: null,
+                                                                 requireExitCodeZero: false,
+                                                                 outputEncoding: null,
+                                                                 killProcessOnCancel: false,
+                                                                 cancellationToken: default(CancellationToken));
+
+                if (exitCode != 0)
+                {
+                    // Rename back to zip
+                    File.Move(taskNugetPath, taskZipPath);
```

Comment:
can we change this to a try/finally and do the move in the finally block? (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/try-finally)

Reply:
actually, can you just call File.Move() before the if?